// @discordx/koa/build/cjs/decorators/classes/DRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DRequest = void 0;
const Method_js_1 = require("./Method.js");
class DRequest extends Method_js_1.Method {
    _api;
    _name;
    _description;
    _type;
    _path;
    _middleware = [];
    get api() {
        return this._api;
    }
    set api(value) {
        this._api = value;
    }
    get type() {
        return this._type;
    }
    get name() {
        return this._name;
    }
    get description() {
        return this._description;
    }
    get path() {
        return this._path;
    }
    get middleware() {
        return this._middleware;
    }
    set middleware(value) {
        this._middleware = value;
    }
    constructor(data) {
        super();
        this._name = data.name;
        this._description = data.description;
        this._type = data.type;
        this._path = data.path;
    }
    setName(value) {
        this._name = value;
    }
    setDescription(value) {
        this._description = value;
    }
    static create(data) {
        return new DRequest(data);
    }
}
exports.DRequest = DRequest;
