// discordx/build/cjs/decorators/decorators/Reaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reaction = void 0;
const index_js_1 = require("../../index.js");
const DReaction_js_1 = require("../classes/DReaction.js");
/**
 * Handle a reaction with a specified emoji (a Unicode string, custom name, or Snowflake)
 *
 * @param options - reaction options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/reaction)
 *
 * @category Decorator
 */
function Reaction(options) {
    return function (target, key) {
        const react = DReaction_js_1.DReaction.create({
            aliases: options?.aliases,
            botIds: options?.botIds,
            description: options?.description,
            directMessage: options?.directMessage,
            emoji: options?.emoji ?? key,
            guilds: options?.guilds,
            partial: options?.partial,
            remove: options?.remove,
        }).decorate(target.constructor, key, target[key]);
        index_js_1.MetadataStorage.instance.addReaction(react);
    };
}
exports.Reaction = Reaction;
