// @discordx/koa/build/cjs/Koa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Koa = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@discordx/di");
const koa_1 = tslib_1.__importDefault(require("koa"));
const metadata_js_1 = require("./logic/metadata.js");
class Koa extends koa_1.default {
    _api;
    _globalMiddlewares;
    constructor(options) {
        super(options);
        this._api = options?.id ?? "api";
        this._globalMiddlewares = options?.globalMiddlewares ?? [];
    }
    // getters
    get api() {
        return this._api;
    }
    set api(value) {
        this._api = value;
    }
    get globalMiddlewares() {
        return this._globalMiddlewares;
    }
    set globalMiddlewares(value) {
        this._globalMiddlewares = value;
    }
    get di() {
        return di_1.DIService.instance;
    }
    get instance() {
        return metadata_js_1.MetadataStorage.instance;
    }
    build() {
        return metadata_js_1.MetadataStorage.instance.build(this);
    }
}
exports.Koa = Koa;
