// @discordx/koa/build/cjs/decorators/decorators/Request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Options = exports.Unlink = exports.Put = exports.Patch = exports.Link = exports.Head = exports.Delete = exports.All = exports.Post = exports.Get = void 0;
const index_js_1 = require("../../index.js");
const metadata_js_1 = require("../../logic/metadata.js");
const DRequest_js_1 = require("../classes/DRequest.js");
function RequestMethod(method, path) {
    return function (target, key) {
        metadata_js_1.MetadataStorage.instance.addRequest(DRequest_js_1.DRequest.create({
            name: key,
            path: path ?? `/${key}`,
            type: method,
        }).decorate(target.constructor, key, target[key]));
    };
}
function Get(path) {
    return RequestMethod(index_js_1.RequestType.Get, path);
}
exports.Get = Get;
function Post(path) {
    return RequestMethod(index_js_1.RequestType.Post, path);
}
exports.Post = Post;
function All(path) {
    return RequestMethod(index_js_1.RequestType.All, path);
}
exports.All = All;
function Delete(path) {
    return RequestMethod(index_js_1.RequestType.Delete, path);
}
exports.Delete = Delete;
function Head(path) {
    return RequestMethod(index_js_1.RequestType.Head, path);
}
exports.Head = Head;
function Link(path) {
    return RequestMethod(index_js_1.RequestType.Link, path);
}
exports.Link = Link;
function Patch(path) {
    return RequestMethod(index_js_1.RequestType.Patch, path);
}
exports.Patch = Patch;
function Put(path) {
    return RequestMethod(index_js_1.RequestType.Put, path);
}
exports.Put = Put;
function Unlink(path) {
    return RequestMethod(index_js_1.RequestType.Unlink, path);
}
exports.Unlink = Unlink;
function Options(path) {
    return RequestMethod(index_js_1.RequestType.Options, path);
}
exports.Options = Options;
