// @discordx/koa/build/esm/decorators/decorators/Request.js
import { RequestType } from "../../index.js";
import { MetadataStorage } from "../../logic/metadata.js";
import { DRequest } from "../classes/DRequest.js";
function RequestMethod(method, path) {
    return function (target, key) {
        MetadataStorage.instance.addRequest(DRequest.create({
            name: key,
            path: path ?? `/${key}`,
            type: method,
        }).decorate(target.constructor, key, target[key]));
    };
}
export function Get(path) {
    return RequestMethod(RequestType.Get, path);
}
export function Post(path) {
    return RequestMethod(RequestType.Post, path);
}
export function All(path) {
    return RequestMethod(RequestType.All, path);
}
export function Delete(path) {
    return RequestMethod(RequestType.Delete, path);
}
export function Head(path) {
    return RequestMethod(RequestType.Head, path);
}
export function Link(path) {
    return RequestMethod(RequestType.Link, path);
}
export function Patch(path) {
    return RequestMethod(RequestType.Patch, path);
}
export function Put(path) {
    return RequestMethod(RequestType.Put, path);
}
export function Unlink(path) {
    return RequestMethod(RequestType.Unlink, path);
}
export function Options(path) {
    return RequestMethod(RequestType.Options, path);
}
